icu4c/source/common/cjmlbe.cpp

```diff
@@ -0,0 +1,462 @@
+    fUnicodeBlock = new UVector32(status);
+    loadMLparams(status);
+}
+
+CjMLBreakEngine::~CjMLBreakEngine() {
+    fUnicodeBlock->removeAllElements();
+    delete fUnicodeBlock;
+    fModel.removeAll();
+    fDigitOrOpenPunctuationOrAlphabetSet.removeAllStrings();
+    fClosePunctuationSet.removeAllStrings();
+}
+
+static const UChar32 INVALID = '|';
+
+static UnicodeString* concatChar(const char* str, UChar32 arr[], UErrorCode& status) {
+    UnicodeString* temp = new UnicodeString(str);
+    int32_t idx = 0;
+    while (arr[idx] != -1) {
+        temp->append(arr[idx++]);
+    }
+    LocalPointer<UnicodeString> result(temp, status);
+    return result.orphan();
+}
+
+static UnicodeString* concatUnicodeString(const char* str1, UnicodeString &str2, UErrorCode& status) {
+    UnicodeString* temp = new UnicodeString(str1);
+    temp->append(str2);
+    LocalPointer<UnicodeString> result(temp, status);
+    return result.orphan();
+}
+
+int32_t CjMLBreakEngine::divideUpDictionaryRange( UText *inText,
+        int32_t rangeStart,
+        int32_t rangeEnd,
+        UVector32 &foundBreaks,
+        UnicodeString &inString,
+        LocalPointer<UVector32> &inputMap,
+        UErrorCode& status) const {
+
+    if (U_FAILURE(status))
+        return 0;
+    if (rangeStart >= rangeEnd) {
+        return 0;
+    }
+
+    UVector32 boundary(inString.countChar32() + 1, status);
+    int32_t numBreaks = 0;
+    UChar32 ch;
+    UnicodeString index;
+    UVector* elementList = new UVector(6, status);
+
+    for (int32_t i = 1; i < inString.length(); i++) {
+        if (elementList->size() > 0) {
+            Element* element = (Element*)elementList->elementAt(0);
+            elementList->removeElementAt(0);
+            delete element;
+            ch = (i + 2) < inString.length() ? inString.char32At(i + 2) : INVALID;
+            index = (ch != INVALID) ? getUnicodeBlockIndex(ch, status) : UnicodeString(INVALID);
+            elementList->addElement(new Element(ch, index), status);
+        }
+        else {
+            initElementList(inString, elementList, status);
+        }
+
+        evaluateBreakpoint (
+            ((Element*)elementList->elementAt(0))->ch,
+            ((Element*)elementList->elementAt(1))->ch,
+            ((Element*)elementList->elementAt(2))->ch,
+            ((Element*)elementList->elementAt(3))->ch,
+            ((Element*)elementList->elementAt(4))->ch,
+            ((Element*)elementList->elementAt(5))->ch,
+            ((Element*)elementList->elementAt(0))->index,
+            ((Element*)elementList->elementAt(1))->index,
+            ((Element*)elementList->elementAt(2))->index,
+            ((Element*)elementList->elementAt(3))->index,
+            ((Element*)elementList->elementAt(4))->index,
+            ((Element*)elementList->elementAt(5))->index,
+            i, numBreaks, boundary, status);
+    }
+    for (int32_t i = elementList->size() - 1; i >= 0; i--) {
```

Comment:
This loop should be avoided by passing a UObjectDeleter in UVector constructor to let the UObjectDeleter delete each element in the destructor instead.

Reply:
Done.